### PR TITLE
Inertia dampeners

### PR DESCRIPTION
**Pull request by neuPanda**
## About the PR

added Inertia dampening states to allow for players to "cruise" while away from the piloting console
added a new station component in shared to allow for the client to be able to identify a station and prevent the user from being able to send an unanchored request for said station.  also use it to intercept and block any request received by the station threw a third party client

as a pilot, on the nav screen i now have an off, dampen, and anchor button
off sets linier and angular dampening to 1
dampen sets it to the configuration in the ship files
anchor sets it to 1 (it takes about 5 seconds to go from top speed to essentially 0)

this will let pilots cruise and not have as many people go to the doctor because they strained their WASD fingers while flying
this will give miners a little bit of safety if while off their ship, their ship is hit by a stray rock.  then if their ship is anchored it will stop its self relatively quickly so as to not be abandon to that rock for the rest of eternity 

## Why / Balance

i think i covered it above, it was changed to save people from finger cramps, and protect miners from those pilots who don't care about safe driving

## How to test

test 1: cant unanchor POI
- put shuttle console on POI's and make sure that the dampener options are not visible

test 2: UI only allows 1 active state
- buy a shuttle, or remove the station component from the POI of the prior test
- re-open the shuttle console
- click the different options, only 1 should be able to be selected at a time

test 3: dampeners off
- using the same shuttle (or toss some thrusters and gyros on your POI if your that dedicated to it)
- set dampers to off
- View Variables the grid server component physics and observe angular and linier dampening = 0
- move your shuttle and let go of controls.  there should be no change in speed once you let go (unless you hit something)

test 4: anchored
- using said shuttle/POI
- set dampeners to anchored
- View the same Variables and observe values now = 1
- test for rapid deceleration by going from top speed to anchored should not instant stop you, but rapidly decelerate

test 5: dampeners on
- using said shuttle or POI (note POI's are going to be a little wonky)
- set dampeners to on
- View the same Variables and observe the values match what is configured in its yaml shuttle component configuration
-- POI's this is normally 999999 
-- Ships this is normally .05
- if on shuttle test for normal deceleration rates we are all use to.  this is the default behavior before any changes were made
- if on POI ensure that this stops at such a speed that you would turn into a puddle on the front wall because going from 20mps to 0 in a millisecond because of 999999 is insane and should crush the vessel but game magics OP 

## Media

image

- [ X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

used the power of the partial classes to extract all of the NF related code for this to the NF directory this should greatly ease any future merge conflicts between this code and upstream

**Changelog**

:cl:
- add: Cruise Control
- add: Parking Breaks
- fix: Fixed fun! (your Prada be driftin')

### CODE REVIEW

**Comment by dvir001** [comment]
``Station`` comp should be added from systems in an automated way (The gamemod file)
In general it should be a more generic name to what it is used for and not added to any map file since it will 100% be forgot from new POI / Maps (Like the 99999 unmovable right now is)

**Review by whatston3 — CHANGES_REQUESTED** [review]
I've put my suggestions up in a branch on my repo, you can see the diff here: https://github.com/neuPanda/frontier-station-14/compare/Inertia-Dampeners...whatston3:frontier-station-14:2024-07-11-Inertia-Dampener-suggestions?expand=1

Things that were changed:
* StationComponent was pushed to the Station itself (populated from station entity definition, see Resources/Prototypes/_NF/Shipyard/base.yml), renamed to StationDampeningComponent.
  * I think this could be hooked into to add the 99999 linear/angular dampening, so that could be removed from the maps themselves, but that was not done in these suggestions.
* Pushed dampening mode into the NavScreenState.  Making a request to change the dampening of a shuttle now sets the dampening for any bound UI.
  * The client's state is now updated through responses from the server, similar to most other BUI.
* The request to change the dampening _will never_ set a shuttle's dampening to 99999.  Ships should not be permanently anchored.
* Changed Anchored to Anchor (consistent tense with Dampen)
* Changed thresholding around the mode calculation - anchored value of 0.1 was a bit arbitrary.
* Added AGPL headers to all _NF files.  Modifications to other files were trivial and left as in-line comments.

Overall, a nice feature, though I think we'll see an increase in ship collisions.  Test suite works as expected, station hiding works fine, and opening multiple simultaneous consoles has changes reflected on both.

**Comment by dvir001** [comment]
> I've put my suggestions up in a branch on my repo, you can see the diff here: https://github.com/neuPanda/frontier-station-14/compare/Inertia-Dampeners...whatston3:frontier-station-14:2024-07-11-Inertia-Dampener-suggestions?expand=1
> 
> Things that were changed:
> 
>     * StationComponent was pushed to the Station itself (populated from station entity definition, see Resources/Prototypes/_NF/Shipyard/base.yml), renamed to StationDampeningComponent.
>       
>       * I think this could be hooked into to add the 99999 linear/angular dampening, so that could be removed from the maps themselves, but that was not done in these suggestions.
> 
>     * Pushed dampening mode into the NavScreenState.  Making a request to change the dampening of a shuttle now sets the dampening for any bound UI.
>       
>       * The client's state is now updated through responses from the server, similar to most other BUI.
> 
>     * The request to change the dampening _will never_ set a shuttle's dampening to 99999.  Ships should not be permanently anchored.
> 
>     * Changed Anchored to Anchor (consistent tense with Dampen), and other "IntertiaDampening" names to be more consistent.
> 
>     * Changed thresholding around the mode calculation - anchored value of 0.1 was a bit arbitrary.
> 
>     * Added AGPL headers to all _NF files.  Modifications to other files were trivial and left as in-line comments.
> 
> 
> Overall, a nice feature, though I think we'll see an increase in ship collisions. Test suite works as expected, station hiding works fine, and opening multiple simultaneous consoles has changes reflected on both.

If this PR still make it possible to fuck with astroids / events / space dungeons this fix will need to be done like the ``Deed`` copy the computer itself confirm its on an actual ship.

**Comment by whatston3** [comment]
> If this PR still make it possible to fuck with astroids / events grids / space dungeons this fix will need to be done like the `Deed` copy the computer itself confirm its on an actual ship.

I'll get that added shortly.

**Comment by whatston3** [comment]
image
Please, do not do this.  You make keeping track of this difficult.

If you want to revert changes, great, do it in the PR.  Do not force push, do not rebase, the changes will get squashed before being merged.

**Comment by whatston3** [comment]
https://github.com/neuPanda/frontier-station-14/compare/Inertia-Dampeners...whatston3:frontier-station-14:2024-07-13-inertia-dampeners-round-2?expand=1

Righto.  New diff.  Issues with buttons not showing up on shuttles should be fixed (I was checking the station entity vs. the grid entity for the ShuttleDeed component - sorry for that), and I haven't seen inconsistencies in the UI between multiple consoles on the same ship.

One caveat: as this exists, this won't play nice if there are multi-grid ships and the ship console is on another grid (unless they all have ship deed or if it's pushed to the shuttle entity)

File: Content.Server/_NF/Shuttles/Systems/ShuttleSystem.cs [hunk 1]
```diff
@@ -29,20 +36,21 @@ private void OnToggleStabilizer(EntityUid uid, ShuttleConsoleComponent component
         {
             InertiaDampeningMode.Off => 0,
             InertiaDampeningMode.Dampen => shuttleComponent.LinearDamping,
-            InertiaDampeningMode.Anchored => 1,
-            _ => 999999,// if it's not the above it's a station, they shouldn't be able to change the dampening, should we alert the admin?
+            InertiaDampeningMode.Anchored => AnchorDampeningStrength,
+            _ => StationaryAnchorStrength,// if it's not the above it's a station, they shouldn't be able to change the dampening, should we alert the admin?
```
Comment (whatston3) [inline Content.Server/_NF/Shuttles/Systems/ShuttleSystem.cs:40]: Shuttles should not be able to set their anchor strength to make them immobile.

Comment (neuPanda) [reply Content.Server/_NF/Shuttles/Systems/ShuttleSystem.cs:40]: they cant.  if the status is not set to one of the 3 above, then it is a hacked client.  and logs will show us that this has happened

**Review by whatston3 — APPROVED** [review]
Made a small bugfix and a few bits of cleanup, nothing huge.  Seems to work alright, and with this, it's fine by me.

Sorry for dragging my feet on this, and thanks again for the feature.

**Comment by dvir001** [comment]
Working fine

**Pull request merged**